README.md
# WaniKani Canvas

The other canvas services had unpleasant limitations so I made this as an experiment. Anyone may view the canvas, but only users with a WaniKani account at level 4 or higher may paint the canvas.

I currently have a server running on AWS.
[Check out the client here!](https://ccookf.com/canvas)

## Server

node, express, socket.io, node-png

## Client

socket.io, jQuery       

### Todo list

* ~~Chunking the canvas~~
* Pretend I'm a CSS wizard


changelog.txt
Version 0.2.2
-----------------------------------------------------------

Client:
* Changed guide mode from [Spacebar] -> [g]
* Changed drag mode to [Spacebar] from [MouseDrag]
* Can paint a continuous stroke with [MouseDown] 
* Can only paint on loaded chunks now

Server:
* Increased the backup threshold for paint operations
* Does an automatic hourly backup
* User count tracking should be more accurate


chat.js
let isInitiated = false;
let chatSocket;

function initChat(io) {
    if (isInitiated) return;
    
    chatSocket = io;
    
    chatSocket.on('message', (data)=>{
        var box = $("#chat-history");
        var st = box.scrollTop();
        var sh = box.prop('scrollHeight');
        var ch = box.prop('clientHeight');
        var isBottomed = (st + ch === sh);
        box.append(formatMessage(data));
        //Autoscroll if the user was at the bottom of chat
        if (isBottomed) {
            box.scrollTop(box.prop('clientHeight'));
        }
    });

    isInitiated = true;
}

function sendMessage() {
    if (isInitiated == false) { console.error("Cannot send message, chat is not initiated!"); return; }
    
    var message = $('#message-box').val();
    if (message == "") return;
    chatSocket.emit('send-message', message, ()=>{
        //clear on server receipt
        $('#message-box').val("");
    });

    console.log("Sending: " + message);
}

//Wipes the existing chat (disconnects, ect) and gets the history
function getChatHistory() {
    chatSocket.emit('get-history', (data)=>{
        if (data == null || data.length == 0) return;
        
        var out = "";
        data.reverse();
        data.forEach((element)=>{
            out = formatMessage(element) + out;
        });

        var box = $("#chat-history");
        box.html(out);

        //Scroll to the bottom of the chat
        box.scrollTop(box.prop('clientHeight'));
    });
}

function formatMessage(data) {
    /**
     * {
     *      username:string,
     *      time:number,
     *      message:string
     * }
     */

    //Simple sanatize of input data
    data.username = data.username.replace("<","&lt;");
    data.message = data.message.replace("<","&lt;");

    //Format the time to HH:MM:SS
    var date = new Date(data.time);
    var time =  "" + padTime((date.getHours()%12)) + ":" 
                + padTime(date.getMinutes()) + ":" 
                + padTime(date.getSeconds());

    var out =   '<div class="chat-message">'
                + '<div class="message-meta">'
                + '<div class="message-user">' + data.username + '</div>'
                + '<div class="message-time">' + time + '</div>'
                + '</div>'
                + '<p class="message-body">' + data.message + '</p></div>';
    return out;
}

function padTime(time) {
    time = time.toString();
    return (time.length < 2) ? '0' + time : time;
}


chunk.js
/**
 * {
 *      isLoaded:bool
 *      column:int
 *      row:int
 * }
 */

var CHUNK_SIZE = 0;     //pixel width of a square chunk
let IMAGE_HEIGHT;   //height of image in pixels
let IMAGE_WIDTH;    //width of image in pixels
let columns;        //number of chunks needed to encompass image width
let rows;           //number of chunks needed to encompass image height
let chunks = [];    //Array of chunks

function initChunks(height, width, size) {
    //Ceil in case image width is not a factor of the chunk size
    columns = Math.ceil(width / size);
    rows = Math.ceil(height / size);
    CHUNK_SIZE = size;
    IMAGE_HEIGHT = height;
    IMAGE_WIDTH = width;

    //Generate array of empty chunks
    for (var i = 0; i < rows; i++) {
        chunks.push([]); //start a new row
        for (var j = 0; j < rows; j++) {
            chunks[i].push({ column: j, row: i, isLoaded: false });
        }
    }

    return { rows: rows, columns: columns }; //For debug
}

/**
 * @description Given the specified bounding box of image space find unloaded chunks
 * @param {number} x Top left of canvas in image space
 * @param {number} y Top left of canvas in image space
 * @param {number} a Bottom right of canvas in image space
 * @param {number} b Bottom right of canvas in image space
 */
function findUnloadedChunks(x, y, a, b, display = false) {

    //I am 90% certain I forgot to account for the negative y space
    //but it somehow worked out. @todo verify this

    //Convert image space coordinates to the chunk space
    var left = Math.floor(x / CHUNK_SIZE);
    var top = Math.floor(y / CHUNK_SIZE);
    var right = Math.floor(a / CHUNK_SIZE) + 1;
    var bottom = Math.floor(b / CHUNK_SIZE) + 1;

    //console.log({left: left, top: top, right: right, bottom: bottom});

    //Clamp the boundaries to the image in chunk space
    left = left < 0 ? 0 : left;
    top = top < 0 ? 0 : top;
    right = right > columns ? columns : right;
    bottom = bottom > rows ? rows : bottom;

    //console.log({left: left, top: top, right: right, bottom: bottom});

    //Search chunks for unloaded members
    var out = [];
    for (var i = top; i < bottom; i++) {
        for (var j = left; j < right; j++) {
            if ((chunks[i][j]).isLoaded == false) out.push(chunks[i][j]);
        }
    }
    if (display) console.log("Found " + out.length + " unloaded chunks.");
    return out;
}

/**
 * @description Determines if the chunk is loaded at the given image space coords
 * @param {number} hor 
 * @param {number} ver 
 */
function isChunkLoaded(hor, ver) {
    var x = Math.floor(hor / CHUNK_SIZE);
    var y = Math.floor(ver / CHUNK_SIZE);

    try {
        if ((chunks[y][x]).isLoaded == false) return false;
        else return true;
    } catch (e) {
        console.log("isChunkLoaded failure:");
        console.log(e);
    }
}

/**
 * @description Sort an array of chunks by distance from a point in image space
 * @param {Array} unsorted Array of image chunks
 * @param {number} x 
 * @param {number} y 
 */
function sortUnloadedChunks(unsorted, x, y) {

    //Convert image space coordinates to the chunk space
    //Adjust for zero indexing
    x = Math.floor(x / CHUNK_SIZE) - 1;
    y = Math.floor(y / CHUNK_SIZE) - 1;

    unsorted.sort((a, b)=>{
        var distA = Math.sqrt((a.column-x)*(a.column-x) + (a.row-y)*(a.row-y));
        var distB = Math.sqrt((b.column-x)*(b.column-x) + (b.row-y)*(b.row-y));
        return distA - distB;
    });
    return unsorted; //for debug
}

/**
 * @Description recursively loads all chunks in the list one at a time
 * @param {Array} list  Sorted array of chunks
 * @param {CanvasRenderingContext2D} context Should be ictx
 */
function loadChunks(list, context, callback) {
    if (list.length == 0) {
        callback(true);
        return;
    }
    var chunk = list.shift(); //When sorted the list is ascending.
    loadChunk(chunk.column, chunk.row, context, (res, time)=>{
        console.log("Chunk (" + chunk.column + ", " + chunk.row + ") loaded: " + res + " in " + time);
        loadChunks(list, context, callback);
    });
}

/**
 * @description Loads the chunk at the specified coordinates in the given canvas context
 * @param {number} column 
 * @param {number} row 
 * @param {CanvasRenderingContext2D} context Should be ictx
 * @param {function} callback Will return the time to load the chunk
 */
function loadChunk(column, row, context, callback) {
    if (chunks[row][column] == true) { console.log("Chunk already loaded!"); return; }
    var start = Date.now();
    socket.emit('get_chunk', { x: column, y: row }, (data)=>{
        var startx = row * CHUNK_SIZE;
        var starty = column * CHUNK_SIZE;

        try{
            var missing = 0;
            for (var i = 0; i < data.length; i++) {
                for (var x = 0; x < data[i].length; x++) {
                    var color = data[i][x]; //Uint8Array [r, g, b]
                    if (!color) { error("Data coordinate missing color data!"); continue; }
                    context.fillStyle = 'rgb(' + color[0] + ',' + color[1] + ',' + color[2] + ')';
                    context.fillRect(startx + x, starty + i, 1, 1);
                }
            }
        } catch (e) {
            console.log("Chunk position: " + row + ", " + column);
            console.log("Failed to load chunk: " + e.message);
            console.log(data);
            if (callback) callback(false, (Date.now() - start) + "ms");
        }
        chunks[row][column].isLoaded = true;
        updateCanvas();
        if (callback) callback(true, (Date.now() - start) + "ms");
    });
}


canvas.js
console.log("canvas script loading...");
if (window.jQuery) console.log("jQuery loaded.");
else error("jQuery has not loaded");

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Initiation //////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

var server = "https://ccookf.com:4242";
//var server = "https://localhost:4242";
var version = "0.2.2";
var isValidUser = false; 

var debug_mode = false;
var scale = [1, 2, 4, 8, 16, 32, 64];

var offset = { x: 0, y: 0 };
var scale_level = 4;

var pos = { x: null, y: null }; //Mouse position in image space as stroke coord
var isMouseDown = false;
var isDragMode = false;
var isGuideMode = false;
var isAltHeld = false;
var isSpaceHeld = false;
var downCoords = { x: 0, y: 0 }; //Coordinates of mousedown event
var mouseOverPixel = { x: 0, y: 0 }; //Current "pixel" under the mouse

var paintColor = { r: 128, g: 255, b: 255 };

// Canvas and image data
var canvas = document.getElementById("paint");
var ctx = canvas.getContext("2d");

//Set initial dimensions of the canvas to match the window
function resizeDisplay() {
	canvas.width = $('.canvas-wrap').width();
	canvas.height = $('.canvas-wrap').height();

	//These seem to reset when the canvas dimensions are adjusted
	//Need to look into why later
	ctx.mozImageSmoothingEnabled = false;
	ctx.webkitImageSmoothingEnabled = false;
	ctx.msImageSmoothingEnabled = false;
	ctx.imageSmoothingEnabled = false;
}
//Listen and resize automatically
$(window).resize(resizeDisplay);
resizeDisplay(); //first time

//Set the window to be unselectable (no highlighting)
window.onload = ()=>{
	//If needed this can be applied per element instead of doc
	document.onselectstart = ()=>{ return false; };
	//document.onmousedown = ()=>{ return false; };
}

var image = document.getElementById("image");
var ictx = image.getContext("2d");

//load the base image onto the image canvas
var img = new Image();

// IO Server
var socket = io(server);
initChat(socket);
socket.on('connect', ()=>{
	console.log("Connected to server successfully");
	getChatHistory();
	socket.emit('get_canvas', (data)=>{

		if (!data) { error("Failed to retreive canvas data."); return; }

		console.log(data);
		if (data.version != version) {
			alert("Client version mismatch. Clear your cache and reload the page. Contact the developer if the problem persists.");
			socket.disconnect();
		}

		//Set the size of the canvas hosting the image to match the data
		image.height = data.height;
		image.width = data.width;
		resetView(); //Sets the canvas offets
		initImage();

		initChunks(data.height, data.width, data.CHUNK_SIZE);
		
		checkUnloadedChunks();
	});

	autoLogin();
});

socket.on('disconnect', ()=>{
	$("#notification").html("Disconnected!");
});

socket.on('user-count', (number)=>{
	if (number == 1) $("#notification").html(number + " user online");
	else $("#notification").html(number + " users online");
});

// User data
var apiKey = localStorage.getItem("wk-api-key");

function autoLogin() {
	if (apiKey == null) {
		$("#login").removeClass("hidden");
	} else {
		socket.emit("verify", apiKey, (res, err)=>{
			if (res == true) {
				isValidUser = true;
				$("#login").html("Logout");
			} else {
				if (confirm("Login failed. Keep API key in storage?") == false) {
					localStorage.removeItem("wk-api-key");
				}
			}
		});
	}
}

$("#login").click(()=>{
	if (!isValidUser) {
		var apiKey = prompt("Please enter your WaniKani API Key");
		if (apiKey == null || apiKey == "") return;
		else {
			socket.emit("verify", apiKey, (res, err)=>{
				if (res == true) {
					localStorage.setItem("wk-api-key", apiKey);
					isValidUser = true;
					$("#login").html("Logout");
				} else {
					alert("Failed to validate user: " + err);
				}
			});
		}
	} else {
		console.log("Logging out");
		socket.emit('logout', (res)=>{
			if (!res) console.error("Problem with logout.");
		});
		localStorage.removeItem("wk-api-key");
		isValidUser = false;
		$("#login").html("Login");
	}
});

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Input Processing ////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

// Zoom Events - mousewheel up/down
canvas.addEventListener("DOMMouseScroll", wheelHandler, false); //firefox
canvas.addEventListener("mousewheel", wheelHandler, false); //everyone else
function wheelHandler(e) {
	var up = e.deltaY > 0 ? false : true;
	if (debug_mode) console.log("Wheel event fired:\t" + e.deltaY + "\t(" + up + ")");
	if (up) zoomIn();
	else zoomOut();
}

// UI events
$("#zoomout").click(zoomOut);
$("#zoomin").click(zoomIn);
$("#reset").click(resetView);
$("#save").click(()=>{
	var link = document.getElementById('save-link');
	link.setAttribute('href', image.toDataURL());
	link.click();
});

var picker = document.getElementById("color-picker");
$("#color-picker").click(()=>{
	var hex = $("#color-picker").val();
	paintColor.r = parseInt(hex.slice(1, 3), 16);
	paintColor.g = parseInt(hex.slice(3, 5), 16);
	paintColor.b = parseInt(hex.slice(5, 7), 16);
});
$("#color-picker").change(()=>{
	var hex = $("#color-picker").val();
	paintColor.r = parseInt(hex.slice(1, 3), 16);
	paintColor.g = parseInt(hex.slice(3, 5), 16);
	paintColor.b = parseInt(hex.slice(5, 7), 16);
});

//Default colors
$("#white").click(()=>{ 	paintColor = { r: 255, 	g: 255, 	b: 255 	};	updateColorPicker();});
$("#red").click(()=>{ 		paintColor = { r: 255, 	g: 0, 		b: 0 	};	updateColorPicker();});
$("#orange").click(()=>{ 	paintColor = { r: 255, 	g: 165, 	b: 0 	};	updateColorPicker();});
$("#peach").click(()=>{ 	paintColor = { r: 255, 	g: 218, 	b: 185 	};	updateColorPicker();});
$("#yellow").click(()=>{ 	paintColor = { r: 255, 	g: 255, 	b: 0 	};	updateColorPicker();});
$("#green").click(()=>{ 	paintColor = { r: 0, 	g: 128, 	b: 0 	};	updateColorPicker();});
$("#blue").click(()=>{ 		paintColor = { r: 0, 	g: 0, 		b: 255 	};	updateColorPicker();});
$("#purple").click(()=>{ 	paintColor = { r: 128, 	g: 0, 		b: 128 	};	updateColorPicker();});
$("#brown").click(()=>{ 	paintColor = { r: 139, 	g: 69, 		b: 19 	};	updateColorPicker();});
$("#black").click(()=>{ 	paintColor = { r: 0, 	g: 0, 		b: 0 	};	updateColorPicker();});

// Chat toggle
$("#toggle-chat").click(()=>{
	var box = $(".main-ui");
	if (box.css("width") == "0px") {
		box.css("width", "30%");
		box.css("min-width", "15em");
	} else {
		box.css("width", "0");
		box.css("min-width", "0");
	}
	resizeDisplay();
});

// Mouse state
canvas.addEventListener("mousedown", (e)=>{
	pos = canvasToImageCoordinate(e.layerX, e.layerY) //update current mouse pos
	if (e.altKey) {
		try {
			var pick = (ctx.getImageData(e.layerX, e.layerY, 1, 1)).data;
			paintColor.r = pick[0];
			paintColor.g = pick[1];
			paintColor.b = pick[2];
			updateColorPicker();
		} catch (e) { error("Failed to pick color: " + e.message); }
	} else {
		isMouseDown = true;
		downCoords.x = e.layerX;
		downCoords.y = e.layerY;

		if (!isDragMode && !isAltHeld) {
			var localCoord = pos;
			//Boundary checking before attempting paint
			if (localCoord.x < 0 || localCoord.y < 0) return;
			if (localCoord.x > image.width || Math.abs(localCoord.y) > image.height) return;
			if (isChunkLoaded(localCoord.x, localCoord.y) == false) return;
			paint(localCoord.x, localCoord.y, ()=>{
				//Do nothing
			});
		}
	}
});
canvas.addEventListener("mouseleave", ()=>{
	isMouseDown = false;
	isDragMode = false;
	updateCanvas(); //Remove the pixel preview, even across windows
});

canvas.addEventListener("mouseup", ()=>{
	isMouseDown = false;
});

canvas.addEventListener("mousemove", mouseMove);

// Keyboard events
window.addEventListener("keyup", (e)=>{
	switch (e.key) {
		case "-":
			zoomOut();
			break;
		case "=":
			zoomIn();
			break;
		case "0":
			resetView();
			break;
		case "g":
			isGuideMode = !isGuideMode;
			console.log("Guide Mode: " + isGuideMode);
			updateCanvas();
			break;
		case " ":
			isDragMode = false;
			break;
		default:
			if (debug_mode) console.log("Unexpected keyboard event: " + e.key);
			return;
	}
});
window.addEventListener("keydown", (e)=>{
	switch (e.key) {
		case " ":
			isDragMode = true;
			break;
		case "Alt":
			updateCanvas(); //clears out temp drawings for pick mode
			isAltHeld = true;
			break;
		default:
			//do nothing
	}
});
window.addEventListener("keyup", (e)=>{
	switch (e.key) {
		case "Alt":
			isAltHeld = false;
			break;
		default:
			//do nothing
	}
});

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Functions ///////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

function updateCanvas(callback) {
	clearCanvas();
	ctx.drawImage(	image, offset.x, offset.y, 
					image.width * scale[scale_level], image.height * scale[scale_level]);

	if (isGuideMode && scale_level >= 3) {
		//determine the total steps on screen
		var zoom = scale[scale_level];
		var steps = { x: canvas.width / zoom, y: canvas.height / zoom };

		//Align the grid the pixels of the image
		var start = {};
		start.x = 0 + Math.floor(offset.x % zoom);
		start.y = 0 + Math.floor(offset.y % zoom);

		//Vertical lines
		ctx.beginPath();
		ctx.strokeStyle = 'rgba(32, 32, 32, 0.2)';
		for (i = 0; i <= steps.x + 1; i++) {
			ctx.moveTo(start.x + zoom * i, start.y);
			ctx.lineTo(start.x + zoom * i, canvas.height);
		}

		//Horizontal lines
		for (i = 0; i <= steps.y + 1; i++) {
			ctx.moveTo(start.x, start.y + zoom * i);
			ctx.lineTo(canvas.width, start.y + zoom * i);
		}
		ctx.closePath();
		ctx.stroke();
	}

	if (callback) { callback(true); }
}

function clearCanvas() {
	ctx.fillStyle = "grey";
	ctx.fillRect(0, 0, canvas.width, canvas.height);
}

function initImage() {
	ictx.fillStyle = "black";
	ictx.fillRect(0, 0, image.width, image.height);
}

function resetView() {
	//Reset the zoom
	scale_level = 0;
	//Get the offsets for the top left corner of the image that would center on canvas
	offset.x = (canvas.width - image.width) / 2;
	offset.y = (canvas.height - image.height) / 2;
	updateCanvas();
}

function zoomIn() {
	if (scale_level < scale.length - 1) {
		
		//Convert the center of the screen to a coordinate in image space
		localCoord = canvasToImageCoordinate(canvas.width/2, canvas.height/2);
		
		//Change the scale
		++scale_level;
		var zoom = scale[scale_level];

		//Revert back from image space coordinates to the center of screen
		offset.x = canvas.width/2 - localCoord.x * zoom;
		offset.y = canvas.height/2 - localCoord.y * zoom;
	}
	updateCanvas();
	checkUnloadedChunks();
}

function zoomOut() {
	if (scale_level > 0) {
		
		localCoord = canvasToImageCoordinate(canvas.width/2, canvas.height/2);
		
		--scale_level;
		var zoom = scale[scale_level];
		offset.x = canvas.width/2 - localCoord.x * zoom;
		offset.y = canvas.height/2 - localCoord.y * zoom;
	}
	updateCanvas();
	checkUnloadedChunks();
}

function canvasToImageCoordinate(x, y) {
	var zoom = scale[scale_level];
	localCoord = { x: 0, y: 0 };
	localCoord.x = Math.floor((x - offset.x) / zoom);
	localCoord.y = Math.floor((y - offset.y) / zoom);

	return localCoord;
}

function imageToCanvasCoordinate(x, y) {
	var zoom = scale[scale_level];
	canvasCoord = { x: 0, y: 0 };
	canvasCoord.x = x * zoom + offset.x;
	canvasCoord.y = y * zoom + offset.y;

	return canvasCoord;
}

function paint(x, y, callback) {
	if (isValidUser == false) return;
	var color = new Uint8Array([paintColor.r, paintColor.g, paintColor.b]);

	var start = { x: pos.x, y: pos.y };

	//Get the difference from the current coordinate and last paint coordinate
	var diff = { x: x - start.x, y: y - start.y };

	//Use the distance as the number of steps when interpolating a stroke
	var steps = Math.sqrt(diff.x * diff.x + diff.y * diff.y);
	steps = steps < 1 ? 1 : steps; //min 1

	for (var i = 0; i < steps; i++) {
		var delta = i / steps;
		var stepx = Math.floor(start.x + diff.x * delta);
		var stepy = Math.floor(start.y + diff.y * delta);

		var out = { color: color, pos: { x: stepx, y: stepy }};
		
		//Don't paint the pixel until the server verifies its fine and announces it
		socket.emit('paint', out);
	}
	callback();
}

//Server announcing the paint from above function
socket.on('paint', (data)=>{
	//console.log(data);
	ictx.fillStyle = 'rgb(' + data.color[0] + ',' + data.color[1] + ',' + data.color[2] + ')';
	ictx.fillRect(data.pos.x, data.pos.y, 1, 1);
	updateCanvas();
});

function mouseMove(e) {
	var x = e.layerX;
	var y = e.layerY;

	//Preview pixel placement. Do not display when using color pick (alt key)
	localCoord = canvasToImageCoordinate(x, y);
	if (!e.altKey && (localCoord.x != mouseOverPixel.x || localCoord.y != mouseOverPixel.y)) {
		//Convert local coordinates to canvas space
		canvasCoord = imageToCanvasCoordinate(localCoord.x, localCoord.y);
		//Draw the image state back onto the canvas, clearing out temp items
		updateCanvas((finished)=>{
			//Draw directly to the main canvas, not the image
			ctx.fillStyle = 'rgb(' + paintColor.r + ',' + paintColor.g + ',' + paintColor.b + ')';
			ctx.fillRect(canvasCoord.x, canvasCoord.y, 1 * scale[scale_level], 1 * scale[scale_level]);
		});
	}

	//Display the mouse coordinates
	centeredCoord = { x: localCoord.x - image.width / 2, y: -1 * (localCoord.y - image.height / 2) };
	$("#coordinates").html("( " + centeredCoord.x + ", " + centeredCoord.y + ")");
	
	if (isMouseDown && isDragMode) {
		offset.x += x - downCoords.x;
		offset.y += y - downCoords.y;

		downCoords.y = y;
		downCoords.x = x;

		updateCanvas();
		checkUnloadedChunks();
	}

	//Paint with strokes
	if (isMouseDown && !isDragMode && !isAltHeld) {
		if (localCoord.x != pos.x || localCoord.y != pos.y) {
			
			//Boundary checking
			if (localCoord.x < 0 || localCoord.y < 0) return;
			if (localCoord.x > image.width || Math.abs(localCoord.y) > image.height) return;
			if (isChunkLoaded(localCoord.x, localCoord.y) == false) return;

			paint(localCoord.x, localCoord.y, ()=>{
				pos = localCoord; //update the last stroke coordinate
			});
		}
	}
}

function colorToHex(color) {
	var out = "#";
	out += decToHexPadded(color.r);
	out += decToHexPadded(color.g);
	out += decToHexPadded(color.b);
	return out;
}

function decToHexPadded(number) {
	number = number.toString(16);
	number = number.length == 1 ? "0" + number : number;
	return number;
}

function updateColorPicker() {
	$("#color-picker").val(colorToHex(paintColor));
}

var isLoadingChunks = false;
function checkUnloadedChunks() {
	var topLeft = canvasToImageCoordinate(0,0);
	var bottomRight = canvasToImageCoordinate(canvas.width, canvas.height);
	var ul = findUnloadedChunks(topLeft.x, topLeft.y, bottomRight.x, bottomRight.y);
	
	if ((ul.length != 0) && (isLoadingChunks == false)) {
		var center = canvasToImageCoordinate(canvas.width/2, canvas.height/2);
		sortUnloadedChunks(ul, center.x, center.y);
		isLoadingChunks = true;
		loadChunks(ul, ictx, (res)=>{
			isLoadingChunks = false;
			console.log("Finished loading chunks.");
		});
	}
}

function error(message) {
	console.log(message);
}


server.js
var fs = require('fs');
var PNG = require('node-png').PNG;
var app = require('express')();
var https = require('https');
var options = {
	key: fs.readFileSync('ccookf_com.key'),
	cert: fs.readFileSync('ccookf_com.crt'),
	ca: fs.readFileSync('ccookf_com.ca-bundle')
}
var httpServer = https.Server(options, app);
var io = require('socket.io')(httpServer);

var port = 4242;

var version = "0.2.2";


var verified = [];
var sockets = [];

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Canvas management ///////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

var CANVAS_WIDTH = 128;
var CANVAS_HEIGHT = 128;
var CHUNK_SIZE = 128;
var canvas = [];
var filename = 'canvas.png';

fs.createReadStream(filename).pipe(new PNG({}))
    .on('error', function(err) {
        console.log("Failed to load png: " + err);
    })
    .on('parsed', function() {
        console.log("Loaded png @" + this.width + "x" + this.height);
        CANVAS_HEIGHT = this.height;
        CANVAS_WIDTH = this.width;

        for (var y = 0; y < this.height; y++) {
            canvas[y] = [];
            for (var x = 0; x < this.width; x++) {
                //No, I don't actually know how the png format is setup
                //I love how much easier the image manip would have been in C++
                var idx = (this.width * y + x) << 2;
                //Javascript is a pain in the butt and Uint8Array([r,g,b]) is probably as efficient
                //as I'm going to get on storage size for colors without doing something annoying
                //so I might as well enjoy MILLIONS OF COLORS LOLOLOLOLOLOL
                //Then someone makes a bot to copy images into the system
                //and it's all like WHOOP WHOOP DATS DA SOUND OF DA POLICE
                //... why am I doing this again?
                canvas[y].push(new Uint8Array([this.data[idx], this.data[idx+1], this.data[idx+2]]));
            }
        }
    });

const BACKUP_THRESHOLD = 250;
var changes = 0;

//Update the canvas and run a backup after BACKUP_THRESHOLD changes occur
function pushPaint(data) {
    //Make sure data is in boundaries
    if (data.pos.x < 0 || data.pos.y < 0 || 
        data.pos.x >= CANVAS_WIDTH || data.pos.y >= CANVAS_HEIGHT) {
            
            return;
    }
    
    changes++;
    canvas[data.pos.y][data.pos.x] = data.color;

    //console.log(data);

    if (changes >= BACKUP_THRESHOLD) {
        //At some point I may want to lock data. Need to look into 
        //possible concurrency issues later
        saveBackup(filename, (res)=>{
            if (true) return;
        });
        changes = 0;
    }
}

//This just saves to a file, but at some point I should shove it out to S3
//And setup numbered output so we can get server history and time lapse
function saveBackup(name, callback) {
    var out = new PNG({width: CANVAS_WIDTH, height: CANVAS_HEIGHT});
    for (var y = 0; y < out.height; y++) {
        for (var x = 0; x < out.width; x++) {
            var idx = (out.width * y + x) << 2;
            out.data[idx] = canvas[y][x][0];
            out.data[idx+1] = canvas[y][x][1];
            out.data[idx+2] = canvas[y][x][2];
            out.data[idx+3] = 255;
        }
    }
    out.pack().pipe(fs.createWriteStream(name));
    callback(true);
}

//Set an hourly time based auto backup.
setInterval(() => {
    saveBackup(filename, (res)=>{
        if (true) console.log("Hourly backup complete. " + Date.now().toString());
        else console.log("Hourly backup failed. " + Date.now().toString());
    })
}, (1000 * 60 * 60));

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Chat ////////////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

var MAX_HISTORY_STORED = 10;
var chatHistory = [{username:"WK CANVAS", time:Date.now(),message:"Server was reset."}];

function updateChat(message, apiKey) {
    
    var out = {};
    console.log("Message from " + apiKey);
    out.username = verified[apiKey].username;
    out.time = Date.now();
    out.message = message;

    chatHistory.push(out);
    if (chatHistory.length > MAX_HISTORY_STORED) chatHistory.shift();
    
    io.emit('message', out);
}

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Server paths ////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////
app.get('/', (req, res)=>{
    res.send('working');
});

// app.get('/canvas', (req, res)=>{
//     res.sendFile();
// });

///////////////////////////////////////////////////////////////////////////////////////////////////
///// Socket events ///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////

io.on('connection', (socket)=>{
    console.log('connection');

    // socket.on('getcanvas', (callback)=>{
    //     callback(canvas);
    // });

    socket.on('get_canvas', (callback)=>{
        var out = { 
            height: canvas.length, 
            width: canvas[0].length, 
            CHUNK_SIZE: CHUNK_SIZE,
            version: version 
        };
        callback(out);
    });

    socket.on('get_chunk', (data, callback)=>{
        //Convert from chunk space to image space
        var x = data.x * CHUNK_SIZE; //row
        var y = data.y * CHUNK_SIZE; //column

        var out = [];

        for (var i = 0; i < CHUNK_SIZE; i++) { //for each row
            out[i] = [];
            for (var j = 0; j < CHUNK_SIZE; j++) { //for each column
                out[i].push(canvas[i + x][j + y]);
            }
        }
        console.log("Served chunk request (" + data.x + ", " + data.y +")");
        callback(out);
    });

    socket.on('paint', (data)=>{
        if (verified[sockets[socket.id]] != null) {
            pushPaint(data);
            io.emit('paint', data);
        }
    });

    socket.on('verify', (apiKey, callback)=>{
        console.log('Received verification request for ' + apiKey);
        if (verified[apiKey] && verified[apiKey].level > 3) {
            console.log('Returning user verified');
            sockets[socket.id] = apiKey;
            callback(true, "You should never see this.");
            return;
        } else {
            console.log(apiKey + ' has not been verified. Pulling WK data.');
            var path = 'https://www.wanikani.com/api/user/'+ apiKey + '/user-information'
            https.get(path, (res)=>{
                var body = '';
                res.on('data', (d)=>{
                    body += d;
                });
                res.on('end', ()=>{
                    console.log("Finished loading WK data for user " + apiKey);
                    try {
                        var user = JSON.parse(body).user_information;
                        verified[apiKey] = user;
                        console.log("Registering user.");
                        if (user == null) {
                            console.log('User data was null.');
                            callback(false, "Failed to get user data.");
                            return;
                        } else if (user.level > 3) {
                            console.log('User is valid.');
                            callback(true, "You should never see this.");
                            sockets[socket.id] = apiKey;
                            return;
                        } else if (user.level < 4) {
                            console.log('User is low level.');
                            callback(false, "Users must be at least level 4.");
                            return;
                        }
                    } catch (e) {
                        console.log(e);
                        callback(false, "Unexpected error.");
                        return;
                    }
                });
            }).on('error', (e)=>{
                console.error(e);
                callback(false, "Something really bad happened.");
                return;
            });
        }
    });

    socket.on('send-message', (data, callback)=>{
        callback(); //Informs client of receipt
        if (!sockets[socket.id]) { console.log("Reject message: not logged in."); return; }
        updateChat(data, sockets[socket.id]);
    });

    socket.on('get-history', (callback)=>{
        callback(chatHistory);
    });

    socket.on('disconnect', ()=>{
        console.log(sockets[socket.id] + " disconnected.");
        delete sockets[socket.id];
    });

    socket.on('logout', (callback)=>{
        console.log(sockets[socket.id] + " disconnected.");
        delete sockets[socket.id];
        callback(true);
    });
});

//Broadcast the number of logged in users periodically
setInterval(()=>{
    io.emit('user-count', Object.keys(sockets).length);
}, 5000);

httpServer.listen(port, ()=>{
    console.log('IO listening on port ' + port);
});
